plugins/android-boot/fu-android-boot-device.c

```diff
@@ -0,0 +1,374 @@
+/*
+ * Copyright (C) 2022 Dylan Van Assche <me@dylanvanassche.be>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <fwupdplugin.h>
+
+#include <fcntl.h>
+#include <string.h>
+
+#include "fu-android-boot-device.h"
+
+#define ANDROID_BOOT_PROC_CMDLINE     "/proc/cmdline"
+#define ANDROID_BOOT_MAX_CMDLINE_SIZE 4096
+#define ANDROID_BOOT_BOOT_SLOT_ARG    "androidboot.slot_suffix"
+#define ANDROID_BOOT_SERIAL_ARG	      "androidboot.serialno"
+#define ANDROID_BOOT_UNKNOWN_VERSION  "0.0.0"
+#define ANDROID_BOOT_SECTOR_SIZE      512
+
+struct _FuAndroidBootDevice {
+	FuUdevDevice parent_instance;
+	const gchar *label;
+	const gchar *boot_slot;
+	const gchar *uuid;
+	const gchar *version;
+	guint64 max_size;
+};
+
+G_DEFINE_TYPE(FuAndroidBootDevice, fu_android_boot_device, FU_TYPE_UDEV_DEVICE)
+
+static void
+fu_android_boot_device_to_string(FuDevice *device, guint idt, GString *str)
+{
+	FuAndroidBootDevice *self = FU_ANDROID_BOOT_DEVICE(device);
+
+	fu_string_append(str, idt, "BootSlot", self->boot_slot);
+	fu_string_append(str, idt, "Label", self->label);
+	fu_string_append(str, idt, "UUID", self->uuid);
+	fu_string_append_kx(str, idt, "MaxSize", self->max_size);
+}
+
+static gboolean
+fu_android_boot_device_probe(FuDevice *device, GError **error)
+{
+	FuAndroidBootDevice *self = FU_ANDROID_BOOT_DEVICE(device);
+	GUdevDevice *udev_device = fu_udev_device_get_dev(FU_UDEV_DEVICE(device));
+	g_autofree gchar *serial = NULL;
+	guint64 sectors = 0;
+	guint64 size = 0;
+	GHashTable *cmdline = NULL;
+
+	/* FuUdevDevice->probe */
+	if (!FU_DEVICE_CLASS(fu_android_boot_device_parent_class)->probe(device, error))
+		return FALSE;
+
+	/* get kernel cmdline */
+	cmdline = fu_kernel_get_cmdline(error);
+
+	/* set the physical ID */
+	if (!fu_udev_device_set_physical_id(FU_UDEV_DEVICE(device), "block", error))
+		return FALSE;
+
+	/* extract boot slot if available */
+	self->boot_slot = g_hash_table_lookup(cmdline, ANDROID_BOOT_BOOT_SLOT_ARG);
+
+	/* extract label and check if it matches boot slot*/
+	if (g_udev_device_has_property(udev_device, "ID_PART_ENTRY_NAME")) {
+		self->label = g_udev_device_get_property(udev_device, "ID_PART_ENTRY_NAME");
+
+		/* Use label as device name */
+		fu_device_set_name(device, self->label);
```

Comment:
like this could be done in fu-udev-device.c.  Then you would use `fu_device_get_name` and no longer need to store to `self->label`.

Reply:
I also thought of this, but the daemon actually does modify the name quite a bit, e.g. it might remove the vendor name -- which isn't what you want if you are going to use it as an instance ID.

Reply:
self->label is only stored for the `to_string`, if we remove it there, it can be a local variable in `_probe`.

Reply:
It's used in `fu_device_add_instance_strsafe(device, "LABEL", self->label);` -- I think it's safer to store tbh.